pkg/kwok/metrics/metrics.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync/atomic"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
+	"sigs.k8s.io/kwok/pkg/kwok/controllers"
+	"sigs.k8s.io/kwok/pkg/kwok/metrics/cel"
+	"sigs.k8s.io/kwok/pkg/log"
+)
+
+// Constants holding metric kinds.
+const (
+	KindGauge     = "gauge"
+	KindHistogram = "histogram"
+	KindCounter   = "counter"
+)
+
+type MetricsUpdateHandler struct {
+	name        string
+	metric      *internalversion.Metric
+	controller  *controllers.Controller
+	environment *cel.Environment
+
+	handler  http.Handler
+	registry *prometheus.Registry
+	updates  []func() error
+}
+
+type MetricsUpdateHandlerConfig struct {
+	NodeName    string
+	Metrics     *internalversion.Metric
+	Controller  *controllers.Controller
+	Environment *cel.Environment
+}
+
+func NewMetricsUpdateHandler(conf MetricsUpdateHandlerConfig) (*MetricsUpdateHandler, error) {
+	registry := prometheus.NewRegistry()
+	handler := promhttp.InstrumentMetricHandler(
+		prometheus.DefaultRegisterer, promhttp.HandlerFor(registry, promhttp.HandlerOpts{Registry: registry}),
+	)
+
+	h := &MetricsUpdateHandler{
+		name:        conf.NodeName,
+		metric:      conf.Metrics,
+		controller:  conf.Controller,
+		environment: conf.Environment,
+		registry:    registry,
+		handler:     handler,
+	}
+	err := h.init()
+	if err != nil {
+		return nil, err
+	}
+	return h, nil
+}
+
+func (h *MetricsUpdateHandler) getNodeInfo(nodeName string) (*controllers.NodeInfo, bool) {
+	nodeInfo, ok := h.controller.GetNode(nodeName)
+	return nodeInfo, ok
+}
+
+func (h *MetricsUpdateHandler) init() error {
+	metric := h.metric
+
+	for _, m := range metric.Spec.Metrics {
+		switch m.Kind {
+		case KindGauge:
+			val := prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Name: m.Name,
+					Help: m.Help,
+					// TODO: ConstLabels
+				},
+			)
+
+			err := h.registry.Register(val)
+			if err != nil {
+				return fmt.Errorf("failed to register gauge %q: %w", m.Name, err)
+			}
+			eval, err := h.environment.Compile(m.Value)
+			if err != nil {
+				return fmt.Errorf("failed to compile metric %q: %w", metric.Name, err)
+			}
+
+			updateFunc := func() error {
+				nodeInfo, ok := h.getNodeInfo(metric.Name)
```

Comment:
Doesn't this mean that the `internalversion.Metric` corresponds to a single `corev1.Node`? 
Follow-up question: to handle labels, we need to register metrics only when the node is in the cache. 
How do you feel about using closure there with sync.Once? Once the `updateFunc` is nil, it won't have a reference to a closure so the closure will be garbage collected, and we don't need to extend any struct
On each call to updateFunc, we call once.Do(register metric with node), and then evaluate a program
We can't really register metric on `NewMetricUpdateHandler` since the metric will be registered before the node is created, and we need the node in order to apply constant labels :relaxed: